kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,700 @@
+	k8svalidation "k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	rcache "github.com/projectcalico/calico/kube-controllers/pkg/cache"
+	client "github.com/projectcalico/calico/libcalico-go/lib/clientv3"
+
+	cnet "github.com/projectcalico/calico/libcalico-go/lib/net"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	uruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	annotationIpv4Pools      = "projectcalico.org/ipv4pools"
+	annotationIpv6Pools      = "projectcalico.org/ipv6pools"
+	annotationLoadBalancerIp = "projectcalico.org/loadBalancerIPs"
+	timer                    = 1 * time.Minute
+)
+
+// loadBalancerController implements the Controller interface for managing Kubernetes services
+// and endpoints, syncing them to the Calico datastore as NetworkSet.
+type loadBalancerController struct {
+	informer        cache.Controller
+	lbResourceCache rcache.ResourceCache
+	calicoClient    client.Interface
+	dataFeed        *node.DataFeed
+	ctx             context.Context
+	cfg             config.LoadBalancerControllerConfig
+	clientSet       *kubernetes.Clientset
+	syncerUpdates   chan interface{}
+	syncStatus      bapi.SyncStatus
+	syncChan        chan interface{}
+	ipamBlocks      map[string]model.KVPair
+	ipPools         map[string]api.IPPool
+}
+
+// NewLoadBalancerController returns a controller which manages Service LoadBalancer objects.
+func NewLoadBalancerController(ctx context.Context, clientset *kubernetes.Clientset, calicoClient client.Interface, cfg config.LoadBalancerControllerConfig) controller.Controller {
+	// set up service informer
+	svcWatcher := cache.NewListWatchFromClient(clientset.CoreV1().RESTClient(), "services", "", fields.Everything())
+
+	listFunc := func() (map[string]interface{}, error) {
+		log.Debugf("Listing profiles from Calico datastore")
+		filteredServices := make(map[string]interface{})
+
+		// Get all profile objects from Calico datastore.
+		serviceList, err := clientset.CoreV1().Services("").List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		// Filter out only services of type LoadBalancer
+		for _, svc := range serviceList.Items {
+			if svc.Spec.Type == v1.ServiceTypeLoadBalancer {
+				key := createHandle(&svc)
+				filteredServices[key] = svc
+			}
+		}
+		log.Debugf("Found %d Service LoadBalancer in Calico datastore", len(filteredServices))
+		return filteredServices, nil
+	}
+
+	cacheArgs := rcache.ResourceCacheArgs{
+		ListFunc:    listFunc,
+		ObjectType:  reflect.TypeOf(&v1.Service{}),
+		LogTypeDesc: "Service",
+	}
+	ccache := rcache.NewResourceCache(cacheArgs)
+
+	svcHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			log.Debugf("Got ADD event for service")
+			svc := obj.(*v1.Service)
+			if isCalicoManagedLoadBalancer(svc, cfg.AssignIPs) {
+				ccache.Set(createHandle(svc), svc)
```

Comment:
Once the ResourceCache stuff is pulled out, we shouldn't be doing any work on any goroutine except for the main worker - all of our event handler should simply queue the received event on a channel for the main worker routine to read and then return.

Much like how you have implemented the `onUpdate` function for Calico resources, we should simply send these service updates through to the main routine via a channel (and probably via a helper structure that contains the update type ADD / DEL / UPD)